features/credentials/provider_credentials_controller.rb

```diff
@@ -0,0 +1,91 @@
+require "set"
+
+require_relative "../../shared/authenticated_controller_base"
+require_relative "../../services/services"
+
+module FastlaneCI
+  # A CRUD controller to manage provider credentials associated with a user. A
+  # `ProviderCredential` is a credential a user may use to access some third-party
+  # provider. For instance, the `GitHubProviderCredential` allows FastlaneCI users
+  # to interact with the GitHub API
+  class ProviderCredentialsController < AuthenticatedControllerBase
+    HOME = "/provider_credentials"
+
+    get HOME do
+      locals = { title: "Provider Credentials" }
+      erb(:provider_credentials, locals: locals, layout: FastlaneCI.default_layout)
+    end
+
+    post "#{HOME}/create" do
+      if valid_params?(params, provider_credential_params) &&
+         valid_user_id?(params[:user_id])
+        Services.provider_credential_service.create_provider_credential!(
+          format_params(params, provider_credential_params)
+        )
+      end
+
+      redirect(HOME)
+    end
+
+    post "#{HOME}/update" do
+      if valid_params?(params, provider_credential_params) &&
+         valid_user_id?(params[:user_id])
+        Services.provider_credential_service.update_provider_credential!(
+          format_params(params, provider_credential_params)
+        )
+      end
+
+      redirect(HOME)
+    end
+
+    private
+
+    #####################################################
+    # @!group Data: View-specific data
+    #####################################################
+
+    # @return [Array[User]]
+    def users
+      Services.user_service.user_data_source.users
+    end
+
+    # Maps users and credentials hash:
+    # { user.id => user.provider_credentials }
+    #
+    # @return [Hash]
+    def user_credentials
+      users
+        .map { |user| [user.id, user.provider_credentials] }
+        .to_h
+    end
+
+    # Empty provider credential
+    #
+    # @return [GitHubProviderCredential]
+    def new_credential
+      @new_credential ||= GitHubProviderCredential.new
+    end
+
+    #####################################################
+    # @!group Locals: View-specific locals
+    #####################################################
+
+    # @return [Set[Symbol]]
+    def provider_credential_params
```

Comment:
Let's make a more descriptive name + comment about how it's used. I think I understand this is used for parameter validation, right? I might even go for `post_parameter_list_for_validation` or something like.